From 1d541ed1018253bb4b9b056f0a9af3d23d39ad0e Mon Sep 17 00:00:00 2001
From: Claudiu Manoil <claudiu.manoil@nxp.com>
Date: Thu, 11 Oct 2018 15:26:27 +0300
Subject: [PATCH 327/706] enetc: Disable control ring when SI resources are
 freed

Signed-off-by: Claudiu Manoil <claudiu.manoil@nxp.com>
(cherry picked from commit e91f1a3a02ceb4ea61adec8ef80df9a4f4674504)
(cherry picked from commit 7fcdf7e3b1d5302f59c74c06ea3aa31845b5171b)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index f1a1edb2d2d2..7c21893f6bd6 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -987,6 +987,11 @@ static void enetc_setup_cbdr(struct enetc_hw *hw, struct enetc_cbdr *cbdr)
 	cbdr->cisr = hw->reg + ENETC_SICBDRCISR;
 }
 
+static void enetc_clear_cbdr(struct enetc_hw *hw)
+{
+	enetc_wr(hw, ENETC_SICBDRMR, 0);
+}
+
 static int enetc_configure_si(struct enetc_ndev_priv *priv)
 {
 	struct enetc_si *si = priv->si;
@@ -1064,6 +1069,7 @@ int enetc_alloc_si_resources(struct enetc_ndev_priv *priv)
 err_config_si:
 	kfree(priv->cls_rules);
 err_alloc_cls:
+	enetc_clear_cbdr(&si->hw);
 	enetc_free_cbdr(priv->dev, &si->cbd_ring);
 err_alloc_cbdr:
 
@@ -1074,6 +1080,7 @@ void enetc_free_si_resources(struct enetc_ndev_priv *priv)
 {
 	struct enetc_si *si = priv->si;
 
+	enetc_clear_cbdr(&si->hw);
 	enetc_free_cbdr(priv->dev, &si->cbd_ring);
 
 	kfree(priv->cls_rules);
-- 
2.17.1

